rohmu/notifier/http.py

```diff
@@ -0,0 +1,98 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
+from .interface import Notifier
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum, unique
+from queue import SimpleQueue
+from typing import Any, Dict, Optional
+
+import requests
+import threading
+
+
+@unique
+class Operation(Enum):
+    UPLOAD = "UPLOAD"
+    DELETE = "DELETE"
+    DELETE_TREE = "DELETE_TREE"
+
+
+def _get_requests_session() -> requests.Session:
+    retry = requests.adapters.Retry(
+        total=3,
+        backoff_factor=0.5,
+        status_forcelist=[429, 500, 502, 503, 504],
+        method_whitelist={"POST"},
+    )
+    adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+    request_session = requests.Session()
+    request_session.mount("http://", adapter)
+    return request_session
+
+
+class HTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._session = _get_requests_session()
+        self._url = url
+
+    def object_created(self, key: str, size: Optional[int]) -> None:
+        self._session.post(
+            self._url,
+            {
+                "op": Operation.UPLOAD.value,
+                "key": key,
+                "size": size,
+                "last_modified": datetime.now(),
+            },
+        )
+
+    def object_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE.value, "key": key})
+
+    def tree_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE_TREE.value, "key": key})
+
+
+@dataclass
+class _HTTPNotifyJob:
+    url: str
+    data: Dict[str, Any]
+
+
+def _background_http_request(queue: "SimpleQueue[_HTTPNotifyJob]") -> None:
```

Comment:
Q: is there a specific reason this a standalone function, and not a method on `BackgroundHTTPNotifier`? almost feels like it was going to be to implement a single thread(pool?) that could handle multiple notifiers?

Reply:
That's true, I've usually designed all my threading stuff by having classes inheriting from `Thread` and have my class basically be the the isolation level/context for the thread. Same result different approach

Reply:
> That's true, I've usually designed all my threading stuff by having classes inheriting from Thread and have my class basically be the the isolation level/context for the thread. Same result different approach

I would agree if that was in a different class, and not on `BackgroundHTTPNotifier`. If you meant something like, `Thread(target=my_logic)` vs `class MyLogic(Thread)`, I guess it is only a matter of taste.